contracts/v2/Outpace.sol

```diff
@@ -0,0 +1,119 @@
+        bytes32[][] proofs;
+        uint256[] amountInTrees;
+        uint256[] amountWithdrawnPerChannel;
+    }
+
+ 	// channelId => channelState
+	mapping (bytes32 => ChannelLibraryV2.State) public states;
+	
+	// withdrawn per channel (channelId => uint)
+	mapping (bytes32 => uint) public remaining;
+	mapping (address => bytes32) public withdrawnPerUser;
+
+	// Events
+	event LogChannelOpen(bytes32 indexed channelId);
+	event LogChannelWithdrawExpired(bytes32 indexed channelId, uint amount);
+	event LogChannelWithdraw(bytes32 indexed channelId, uint amount);
+
+	// All functions are external
+	function channelOpen(ChannelLibraryV2.Channel calldata channel)
+		external
+	{
+		bytes32 channelId = ChannelLibraryV2.hash(channel);
+		require(states[channelId] == ChannelLibraryV2.State.Unknown, "INVALID_STATE");
+		require(msg.sender == channel.creator, "INVALID_CREATOR");
+		require(ChannelLibraryV2.isValid(channel, now), "INVALID_CHANNEL");
+		
+		states[channelId] = ChannelLibraryV2.State.Active;
+        remaining[channelId] = channel.tokenAmount;
+
+		SafeERC20.transferFrom(channel.tokenAddr, msg.sender, address(this), channel.tokenAmount);
+
+		emit LogChannelOpen(channelId);
+	}
+
+	function channelWithdrawExpired(ChannelLibraryV2.Channel calldata channel)
+		external
+	{
+		bytes32 channelId = ChannelLibraryV2.hash(channel);
+		require(states[channelId] == ChannelLibraryV2.State.Active, "INVALID_STATE");
+		require(now > channel.validUntil, "NOT_EXPIRED");
+		require(msg.sender == channel.creator, "INVALID_CREATOR");
+		
+		uint toWithdraw = remaining[channelId];
+
+		// NOTE: we will not update withdrawn, since a WithdrawExpired does not count towards normal withdrawals
+		states[channelId] = ChannelLibraryV2.State.Expired;
+		
+		SafeERC20.transfer(channel.tokenAddr, msg.sender, toWithdraw);
+
+		emit LogChannelWithdrawExpired(channelId, toWithdraw);
+	}
+
+    function channelWithdrawBulk(BulkWithdraw calldata bulkWithdraw)
+        external
+    {
+
+        // validate withdrawn hash
+        if(bulkWithdraw.amountWithdrawnPerChannel.length > 0 ) {
+            require(bulkWithdraw.amountWithdrawnPerChannel.computeRoot(msg.sender) == withdrawnPerUser[msg.sender], 'INVALID_WITHDRAW_DATA');
+        } else {
+            require(withdrawnPerUser[msg.sender] == bytes32(0), 'INVALID_WTHDRAWAL_DATA');
+        }
+
+        uint256 currentTotalAmountToWithdraw = 0;
+        
+        for(uint i = 0; i < bulkWithdraw.channels.length; i++) {
+            ChannelLibraryV2.Channel calldata channel = bulkWithdraw.channels[i];
+            bytes32 stateRoot = bulkWithdraw.stateRoots[i];
+            uint amountInTree = bulkWithdraw.amountInTrees[i];
+            bytes32[3][] calldata signature = bulkWithdraw.signatures[i];
+
+            bytes32 channelId = ChannelLibraryV2.hash(channel);
+            require(states[channelId] == ChannelLibraryV2.State.Active, "INVALID_STATE");
+            require(now <= channel.validUntil, "EXPIRED");
```

Comment:
we may be getting rid of that check: this is something that's described in the AIP